### ISSUE TITLE
LSTM learning with correct(target) input provided to the decoder

### ISSUE THREAD

**Issue by yurkor** [fchollet/keras#3263]
I have feel that current keras examples LSTM of seq to seq training is suboptimal, correct me if I wrong. 
It works for simple tasks like https://github.com/fchollet/keras/blob/master/examples/addition_rnn.py.
But not for Neural conversation model https://arxiv.org/pdf/1506.05869.pdf and machine translation. 
Most of the results of seq-seq model returns common repetitive(like "the the the a" in dialog generation) words and reason I believe in that LSTM accumulate errors when using generated output of previous state as input for next state. It would be nice to have ability to use target output as input for next step.
Tensor-flow does this in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/seq2seq.py with feed_previous.
Code in keras is more general but I would like to get some help on ideas how could this be implemented in keras. 

I think may be special LSTM version which combines softmax output and uses K.use_in_train(target_embedding, prev_state)

Or may be have additional input with shifted target (y_in[0:] = y[1:]) to LSTM concatenated with other inputs but then its unclear how to test this model.

**Comment by mbollmann** [comment]
> Or may be have additional input with shifted target (y_in[0:] = y[1:]) to LSTM concatenated with other inputs but then its unclear how to test this model.

FWIW, I'm currently doing exactly this. Since I'm mostly interested in full-sequence accuracy, testing works basically the same as training (i.e., if all predictions are correct the decoder input will be identical to the shifted target, so you can just use that).

For actual prediction, I'm only aware of makeshift solutions (feeding sequentially longer inputs to the decoder, each time only looking at the output prediction at timestep *i* to concat this to the input for timestep *i+1*, and so on...). If it would be possible to implement this in a more elegant and efficient way, this would be a huge help.

**Comment by yurkor** [comment]
How much does it help with prediction accuracy on  makeshift solutions ? Most of Theano based seq2seq models aren't successful like https://github.com/nicolas-ivanov/seq2seq_chatbot_links I wander if that was an issue.

**Comment by booble123** [comment]
There is implementation of seq2seq autoencoder in Torch7 that does exactly this:
- feeds target values (shifted to the right by 1 with 0.0 prepended) as second input to decoder
- takes t-1 output from decoder and uses as additional input at time step t
https://github.com/Kaixhin/Autoencoders/blob/master/models/Seq2SeqAE.lua

As you can see there is a big "if" block that does different things based on whether it's in training mode or not.

AFAIK you can't do that in Keras, but following should work:

1) Take targets, prepend 0.0 on the left and feed to decoder as second input; you need to prepend 0.0 since decoder takes its own output at t-1 at time step t, so at t=0 you need to provide some "zero state" value

2) It's going to be a mess at inference time. You'll have to feed your sequence step by step.
Output of decoder at step t-1 should be used as second input at time step t.

I'm going to try exactly that in coming days.

**Comment by mbollmann** [comment]
> 1) Take targets, prepend 0.0 on the left and feed to decoder as second input; you need to prepend 0.0 since decoder takes its own output at t-1 at time step t, so at t=0 you need to provide some "zero state" value

FWIW, I actually prepend a special "start symbol" for that purpose.  I haven't tried using zeroes, don't know if it makes a difference.

> 2) It's going to be a mess at inference time. You'll have to feed your sequence step by step.
Output of decoder at step t-1 should be used as second input at time step t.

Yes. I outlined my thoughts on how to do this in [a different issue](https://github.com/fchollet/keras/issues/2968#issuecomment-226727979), if you haven't seen it yet. Maybe it helps. I'd be interested to see if you come up with something!

**Comment by erickrf** [comment]
By feeding the correct input, you mean feeding the word embedding of the expected word at time t?

**Comment by ShivaliGoel** [comment]
@erickrf yes ! Please suggest how to feed the word embedding of the expected word at time t?

**Comment by erickrf** [comment]
@ShivaliGoel I don't know. In the end I gave up using keras for autoencoders and used tensorflow.
